core/sql/executor/ExHbaseIUD.cpp

```diff
@@ -4024,36 +4117,79 @@ ExWorkProcRetcode ExHbaseAccessSQRowsetTcb::work()
 	case PROCESS_UPDATE:
 	case PROCESS_UPDATE_AND_CLOSE:
 	  {
-            short numRowsInBuffer = patchDirectRowBuffers();
+            numRowsInVsbbBuffer_ = patchDirectRowBuffers();
 
 	    retcode = ehi_->insertRows(table_,
 				       hbaseAccessTdb().getRowIDLen(),
                                        rowIDs_,
                                        rows_,
                                        hbaseAccessTdb().useHbaseXn(),
-				       -1);
-
+				       -1,
+				       hbaseAccessTdb().asyncOperations());
             currRowNum_ = 0;	    
 	    if (setupError(retcode, "ExpHbaseInterface::insertRows"))
 	      {
 		step_ = HANDLE_ERROR;
 		break;
 	      }
-
-            if (getHbaseAccessStats())
-	    {
+            if (hbaseAccessTdb().asyncOperations()) {
+               lastHandledStep_ = step_;
+               step_ = ASYNC_INSERT_COMPLETE;
+               break;
+            }
+            if (getHbaseAccessStats()) {
 	      getHbaseAccessStats()->lobStats()->numReadReqs++;
-	      getHbaseAccessStats()->incUsedRows(numRowsInBuffer);
-
+	      getHbaseAccessStats()->incUsedRows(numRowsInVsbbBuffer_);
 	    }
-
 	    if (step_ == PROCESS_UPDATE_AND_CLOSE)
 	      step_ = RS_CLOSE;
 	    else
 	      step_ = DONE;
 	  }
 	  break;
-
+      case ASYNC_INSERT_COMPLETE:
+         {
+            if (resultArray_  == NULL)
+                resultArray_ = new (getHeap()) NABoolean[hbaseAccessTdb().getHbaseRowsetVsbbSize()];
+            Int32 timeout;
+            if (asyncCompleteRetryCount_ < 10)
+               timeout = -1;
+            else {
+               asyncOperationTimeout_ = asyncOperationTimeout_ * 2;
+               timeout = asyncOperationTimeout_;
+            }
+            retcode = ehi_->completeAsyncOperation(timeout, resultArray_, numRowsInVsbbBuffer_);
```

Comment:
Later it might be good to refactor code from lines 4151 to 4170 (or maybe some other subset) into a separate method so that we don't have the same code block in multiple places (only if it is good design).